lib/Sema/CSFix.cpp

```diff
@@ -1934,7 +1934,7 @@ bool AllowNonClassTypeToConvertToAnyObject::diagnose(const Solution &solution,
                                                      bool asNote) const {
   auto *locator = getLocator();
 
-  if (locator->isForContextualType()) {
+  if (locator->isForContextualType() || locator->isForTupleElement()) {
```

Comment:
Instead of doing this let's just add a new `FailureDiagnostic` (something like `InvalidNonClassTypeToAnyObjectConversion`) and move `if (locator->isForContextualType()` and `if (locator->isLastElement<LocatorPathElt::ApplyArgToParam>())` to its `::diagnoseAsError()` method, add new custom cases for collections and a fallback to the generic `value of type 'String' expected to be instance of class or class-constrained type`. This way we don't have to add `isForTupleElement()` and assume that it would always have a contextual type.